In [ ]:
!pip install neuralforecast

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.6/170.6 kB 4.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 776.9/776.9 kB 11.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.5/62.5 MB 13.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 409.6/409.6 kB 45.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 805.2/805.2 kB 53.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.7/101.7 kB 11.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 226.8/226.8 kB 20.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.6/78.6 kB 10.4 MB/s eta 0:00:00


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import math
import numpy as np
import torch
from typing import Optional, Union, Tuple


import torch.nn.functional as F
from torch.distributions import Distribution
from torch.distributions import Bernoulli, Normal, StudentT, Poisson, NegativeBinomial

from torch.distributions import constraints


from neuralforecast import NeuralForecast
from neuralforecast.models import NHITS
from neuralforecast.tsdataset import TimeSeriesDataset
from neuralforecast.utils import AirPassengers, AirPassengersPanel, AirPassengersStatic
from neuralforecast.losses.numpy import rmse, mape
from pandas.tseries.holiday import USFederalHolidayCalendar
from pandas.tseries.offsets import CustomBusinessDay
from neuralforecast.losses.pytorch import MQLoss, DistributionLoss, MSE, MAE

In [ ]:
from numpy.random import seed
from random import randrange

In [ ]:
df = pd.read_excel("SP100_vol.xlsx", index_col="Date",parse_dates=True)
columns=df.columns
df_static=pd.read_excel("Static_df.xlsx")
n_series=len(df.columns)
train = df.iloc[:int(len(df['NKE'])*0.7)]
validation_length=int(len(df['NKE'])*0.7)-int(len(df['NKE'])*0.5)
train=train.unstack().reset_index(name="Value").rename(columns={"level_0":"unique_id", "Date":"ds", "Value":"y"})

In [ ]:
n_inputs = [3,5,10,15,21,42,84]
mlp_units = [[[712, 712], [712, 712]],[[512, 512], [512, 512]],[[250, 250], [250, 250]],[[100, 100], [100, 100]]]
epochs=[50,100,150,250,350,450,550,650,750]
learning_rate=[0.0005,0.0001,0.00005,0.00001]
num_lr_decays=[5,3,2,1]
scaler_type=["robust","standard",'minmax']
losses=[MSE(),MAE(),MQLoss(level=[80, 90]),DistributionLoss(distribution='StudentT', level=[80, 90])]
n_blocks=[[1, 1],[2, 2],[3, 3],[5, 5]]
dropout_prob_theta=[0, 0.2, 0.3, 0.4]
n_pool_kernel_size=[[2, 2, 1], [4, 2, 1], [4, 4, 2]]
pooling_mode=["MaxPool1d", "AvgPool1d"]
interpolation_mode=['linear', 'nearest']

# Hyperparameter Search

In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(4)
  m=randrange(4)
  a=randrange(4)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(4)
  a2=randrange(3)
  a3=randrange(2)
  a4=randrange(2)
  model = NHITS(h=1, input_size=n_inputs[i],
                loss=losses[m],
                scaler_type=scaler_type[o],
                learning_rate=learning_rate[n],
                stack_types=['identity','identity'],
                n_blocks=n_blocks[a],
                mlp_units=mlp_units[h],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[n],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=epochs[k],
                dropout_prob_theta = dropout_prob_theta[a1],
                n_pool_kernel_size=n_pool_kernel_size[a2],
                pooling_mode=pooling_mode[a3],
                interpolation_mode=interpolation_mode[a4],
                random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | MLP units: {mlp_units[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | Interpolation: {interpolation_mode[a4]}")
  print(f"Learning Rate: {n} | Block: {n_blocks[a]} | Dropout: {dropout_prob_theta[a1]} | Pooling: {pooling_mode[a3]} | Kernel: {n_pool_kernel_size[a2]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 59107986


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 73661452


Epoch: 250 | Inputs: 42 | MLP units: [[512, 512], [512, 512]] | Loss: DistributionLoss() | Scaler Type: standard | Interpolation: linear
Learning Rate: 1 | Block: [3, 3] | Dropout: 0.2 | Pooling: AvgPool1d | Kernel: [2, 2, 1]
RMSE: 0.002744229734720844 | QLIKE: 0.027375017617260893 | MAE: 0.0016164945297726194 | MAPE: 0.16202469672416023


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 82950131


Epoch: 450 | Inputs: 84 | MLP units: [[250, 250], [250, 250]] | Loss: DistributionLoss() | Scaler Type: robust | Interpolation: nearest
Learning Rate: 3 | Block: [2, 2] | Dropout: 0 | Pooling: MaxPool1d | Kernel: [4, 2, 1]
RMSE: 0.0028697431755798047 | QLIKE: 0.026657627145320387 | MAE: 0.0016615304383306917 | MAPE: 0.17194527242487467


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 89846843


Epoch: 50 | Inputs: 15 | MLP units: [[100, 100], [100, 100]] | Loss: MAE() | Scaler Type: minmax | Interpolation: nearest
Learning Rate: 3 | Block: [1, 1] | Dropout: 0.2 | Pooling: MaxPool1d | Kernel: [4, 2, 1]
RMSE: 0.00281794306951875 | QLIKE: 0.02696654577131562 | MAE: 0.0014689300133735087 | MAPE: 0.14817414569933018


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 112908824


Epoch: 250 | Inputs: 21 | MLP units: [[250, 250], [250, 250]] | Loss: MQLoss() | Scaler Type: minmax | Interpolation: linear
Learning Rate: 3 | Block: [3, 3] | Dropout: 0.2 | Pooling: AvgPool1d | Kernel: [2, 2, 1]
RMSE: 0.0027869500573040567 | QLIKE: 0.028187314963111436 | MAE: 0.0014603015350447242 | MAPE: 0.14614792194397855


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 100 | Inputs: 3 | MLP units: [[100, 100], [100, 100]] | Loss: DistributionLoss() | Scaler Type: standard | Interpolation: nearest
Learning Rate: 1 | Block: [3, 3] | Dropout: 0.4 | Pooling: MaxPool1d | Kernel: [2, 2, 1]
RMSE: 0.002764413591381078 | QLIKE: 0.02829557102123029 | MAE: 0.001454058037620249 | MAPE: 0.14401676939646418


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(4)
  m=randrange(4)
  a=randrange(4)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(4)
  a2=randrange(3)
  a3=randrange(2)
  a4=randrange(2)
  model = NHITS(h=1, input_size=n_inputs[i],
                loss=losses[m],
                scaler_type=scaler_type[o],
                learning_rate=learning_rate[n],
                stack_types=['identity','identity'],
                n_blocks=n_blocks[a],
                mlp_units=mlp_units[h],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[n],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=epochs[k],
                dropout_prob_theta = dropout_prob_theta[a1],
                n_pool_kernel_size=n_pool_kernel_size[a2],
                pooling_mode=pooling_mode[a3],
                interpolation_mode=interpolation_mode[a4],
                random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | MLP units: {mlp_units[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | Interpolation: {interpolation_mode[a4]}")
  print(f"Learning Rate: {n} | Block: {n_blocks[a]} | Dropout: {dropout_prob_theta[a1]} | Pooling: {pooling_mode[a3]} | Kernel: {n_pool_kernel_size[a2]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 69220072


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 86727879


Epoch: 250 | Inputs: 42 | MLP units: [[100, 100], [100, 100]] | Loss: MQLoss() | Scaler Type: robust | Interpolation: linear
Learning Rate: 1 | Block: [5, 5] | Dropout: 0.2 | Pooling: AvgPool1d | Kernel: [2, 2, 1]
RMSE: 0.0027467581747816663 | QLIKE: 0.027201486579215273 | MAE: 0.001482398413567958 | MAPE: 0.14681356366492576


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 34834070


Epoch: 650 | Inputs: 84 | MLP units: [[250, 250], [250, 250]] | Loss: MAE() | Scaler Type: standard | Interpolation: linear
Learning Rate: 1 | Block: [1, 1] | Dropout: 0 | Pooling: MaxPool1d | Kernel: [2, 2, 1]
RMSE: 0.0027230473381928077 | QLIKE: 0.027637475162554707 | MAE: 0.0016440202589433964 | MAPE: 0.15725013851484274


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 29300786


Epoch: 100 | Inputs: 15 | MLP units: [[250, 250], [250, 250]] | Loss: DistributionLoss() | Scaler Type: minmax | Interpolation: linear
Learning Rate: 2 | Block: [5, 5] | Dropout: 0.4 | Pooling: AvgPool1d | Kernel: [4, 4, 2]
RMSE: 0.0027912470152798444 | QLIKE: 0.027647276903758906 | MAE: 0.0014714429676830698 | MAPE: 0.14734860313428055


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 34140381


Epoch: 150 | Inputs: 5 | MLP units: [[100, 100], [100, 100]] | Loss: MQLoss() | Scaler Type: minmax | Interpolation: nearest
Learning Rate: 0 | Block: [2, 2] | Dropout: 0.4 | Pooling: MaxPool1d | Kernel: [4, 2, 1]
RMSE: 0.0027391878979135392 | QLIKE: 0.026681482808778273 | MAE: 0.0014566612114571018 | MAPE: 0.14581243939386965


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 50 | Inputs: 5 | MLP units: [[100, 100], [100, 100]] | Loss: MAE() | Scaler Type: standard | Interpolation: linear
Learning Rate: 2 | Block: [3, 3] | Dropout: 0.3 | Pooling: MaxPool1d | Kernel: [4, 4, 2]
RMSE: 0.002781363603031141 | QLIKE: 0.02815873350633971 | MAE: 0.001446136466298664 | MAPE: 0.14402319393307303


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(4)
  m=randrange(4)
  a=randrange(4)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(4)
  a2=randrange(3)
  a3=randrange(2)
  a4=randrange(2)
  model = NHITS(h=1, input_size=n_inputs[i],
                loss=losses[m],
                scaler_type=scaler_type[o],
                learning_rate=learning_rate[n],
                stack_types=['identity','identity'],
                n_blocks=n_blocks[a],
                mlp_units=mlp_units[h],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[n],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=epochs[k],
                dropout_prob_theta = dropout_prob_theta[a1],
                n_pool_kernel_size=n_pool_kernel_size[a2],
                pooling_mode=pooling_mode[a3],
                interpolation_mode=interpolation_mode[a4],
                random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | MLP units: {mlp_units[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | Interpolation: {interpolation_mode[a4]}")
  print(f"Learning Rate: {n} | Block: {n_blocks[a]} | Dropout: {dropout_prob_theta[a1]} | Pooling: {pooling_mode[a3]} | Kernel: {n_pool_kernel_size[a2]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 6589365


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 5055240


Epoch: 150 | Inputs: 42 | MLP units: [[100, 100], [100, 100]] | Loss: MSE() | Scaler Type: robust | Interpolation: nearest
Learning Rate: 1 | Block: [3, 3] | Dropout: 0.2 | Pooling: AvgPool1d | Kernel: [4, 2, 1]
RMSE: 0.002771093489896942 | QLIKE: 0.026570721580167617 | MAE: 0.0014931369837498372 | MAPE: 0.1511215962254253


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 82444203


Epoch: 150 | Inputs: 10 | MLP units: [[100, 100], [100, 100]] | Loss: MSE() | Scaler Type: robust | Interpolation: nearest
Learning Rate: 3 | Block: [5, 5] | Dropout: 0 | Pooling: AvgPool1d | Kernel: [4, 2, 1]
RMSE: 0.0027818406621551227 | QLIKE: 0.02663499835506535 | MAE: 0.0014506427422268106 | MAPE: 0.14677144058546246


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 100584735


Epoch: 750 | Inputs: 15 | MLP units: [[250, 250], [250, 250]] | Loss: MSE() | Scaler Type: robust | Interpolation: nearest
Learning Rate: 0 | Block: [1, 1] | Dropout: 0.2 | Pooling: AvgPool1d | Kernel: [4, 2, 1]
RMSE: 0.00279533681953325 | QLIKE: 0.029262431034538683 | MAE: 0.0016978148101125356 | MAPE: 0.1697482373391697


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 59393529


Epoch: 350 | Inputs: 10 | MLP units: [[512, 512], [512, 512]] | Loss: MSE() | Scaler Type: minmax | Interpolation: linear
Learning Rate: 1 | Block: [5, 5] | Dropout: 0 | Pooling: MaxPool1d | Kernel: [4, 4, 2]
RMSE: 0.0026958808611141196 | QLIKE: 0.02544367298506698 | MAE: 0.0017175365179240164 | MAPE: 0.1758057730138017


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 450 | Inputs: 21 | MLP units: [[512, 512], [512, 512]] | Loss: DistributionLoss() | Scaler Type: minmax | Interpolation: linear
Learning Rate: 1 | Block: [1, 1] | Dropout: 0.2 | Pooling: MaxPool1d | Kernel: [2, 2, 1]
RMSE: 0.0027040353019622144 | QLIKE: 0.02712200475317952 | MAE: 0.001631905411313709 | MAPE: 0.15554612373141502


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(4)
  m=randrange(4)
  a=randrange(4)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(4)
  a2=randrange(3)
  a3=randrange(2)
  a4=randrange(2)
  model = NHITS(h=1, input_size=n_inputs[i],
                loss=losses[m],
                scaler_type=scaler_type[o],
                learning_rate=learning_rate[n],
                stack_types=['identity','identity'],
                n_blocks=n_blocks[a],
                mlp_units=mlp_units[h],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[n],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=epochs[k],
                dropout_prob_theta = dropout_prob_theta[a1],
                n_pool_kernel_size=n_pool_kernel_size[a2],
                pooling_mode=pooling_mode[a3],
                interpolation_mode=interpolation_mode[a4],
                random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | MLP units: {mlp_units[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | Interpolation: {interpolation_mode[a4]}")
  print(f"Learning Rate: {n} | Block: {n_blocks[a]} | Dropout: {dropout_prob_theta[a1]} | Pooling: {pooling_mode[a3]} | Kernel: {n_pool_kernel_size[a2]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 77483894


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 40346780


Epoch: 150 | Inputs: 21 | MLP units: [[712, 712], [712, 712]] | Loss: DistributionLoss() | Scaler Type: standard | Interpolation: nearest
Learning Rate: 0 | Block: [2, 2] | Dropout: 0 | Pooling: MaxPool1d | Kernel: [4, 2, 1]
RMSE: 0.00261795408412616 | QLIKE: 0.024710279430817574 | MAE: 0.0016030544887469216 | MAPE: 0.15921628282272618


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 42295783


Epoch: 50 | Inputs: 10 | MLP units: [[250, 250], [250, 250]] | Loss: MSE() | Scaler Type: robust | Interpolation: linear
Learning Rate: 2 | Block: [2, 2] | Dropout: 0.3 | Pooling: MaxPool1d | Kernel: [4, 2, 1]
RMSE: 0.0027490120379897735 | QLIKE: 0.027449293329355317 | MAE: 0.0014388418153172416 | MAPE: 0.14318309819772287


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 25048609


Epoch: 100 | Inputs: 42 | MLP units: [[250, 250], [250, 250]] | Loss: DistributionLoss() | Scaler Type: robust | Interpolation: linear
Learning Rate: 1 | Block: [1, 1] | Dropout: 0.2 | Pooling: AvgPool1d | Kernel: [4, 4, 2]
RMSE: 0.002821529313126574 | QLIKE: 0.026576617593109284 | MAE: 0.0015293773987400277 | MAPE: 0.15545278981943006


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 72701007


Epoch: 250 | Inputs: 3 | MLP units: [[250, 250], [250, 250]] | Loss: DistributionLoss() | Scaler Type: minmax | Interpolation: nearest
Learning Rate: 2 | Block: [5, 5] | Dropout: 0.3 | Pooling: MaxPool1d | Kernel: [2, 2, 1]
RMSE: 0.0027326166917372675 | QLIKE: 0.02678885251127744 | MAE: 0.001519869907308776 | MAPE: 0.15138392499973907


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 350 | Inputs: 15 | MLP units: [[512, 512], [512, 512]] | Loss: DistributionLoss() | Scaler Type: minmax | Interpolation: linear
Learning Rate: 0 | Block: [2, 2] | Dropout: 0 | Pooling: AvgPool1d | Kernel: [4, 4, 2]
RMSE: 0.002765160590040289 | QLIKE: 0.028323861640038153 | MAE: 0.0015386054797144675 | MAPE: 0.15260393735960462


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(4)
  m=randrange(4)
  a=randrange(4)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(4)
  a2=randrange(3)
  a3=randrange(2)
  a4=randrange(2)
  model = NHITS(h=1, input_size=n_inputs[i],
                loss=losses[m],
                scaler_type=scaler_type[o],
                learning_rate=learning_rate[n],
                stack_types=['identity','identity'],
                n_blocks=n_blocks[a],
                mlp_units=mlp_units[h],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[n],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=epochs[k],
                dropout_prob_theta = dropout_prob_theta[a1],
                n_pool_kernel_size=n_pool_kernel_size[a2],
                pooling_mode=pooling_mode[a3],
                interpolation_mode=interpolation_mode[a4],
                random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | MLP units: {mlp_units[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | Interpolation: {interpolation_mode[a4]}")
  print(f"Learning Rate: {n} | Block: {n_blocks[a]} | Dropout: {dropout_prob_theta[a1]} | Pooling: {pooling_mode[a3]} | Kernel: {n_pool_kernel_size[a2]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 44502625


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 40934026


Epoch: 250 | Inputs: 10 | MLP units: [[512, 512], [512, 512]] | Loss: DistributionLoss() | Scaler Type: robust | Interpolation: linear
Learning Rate: 0 | Block: [2, 2] | Dropout: 0 | Pooling: AvgPool1d | Kernel: [4, 2, 1]
RMSE: 0.0028067660617963415 | QLIKE: 0.027036301139216425 | MAE: 0.00165587288399703 | MAPE: 0.1713017935233765


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 104711776


Epoch: 650 | Inputs: 42 | MLP units: [[512, 512], [512, 512]] | Loss: MSE() | Scaler Type: robust | Interpolation: nearest
Learning Rate: 1 | Block: [2, 2] | Dropout: 0.2 | Pooling: MaxPool1d | Kernel: [4, 2, 1]
RMSE: 0.002634925480244614 | QLIKE: 0.02482087674308563 | MAE: 0.0016500003385030387 | MAPE: 0.16646681663884497


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 75407892


Epoch: 100 | Inputs: 3 | MLP units: [[100, 100], [100, 100]] | Loss: MAE() | Scaler Type: minmax | Interpolation: linear
Learning Rate: 2 | Block: [3, 3] | Dropout: 0.4 | Pooling: MaxPool1d | Kernel: [2, 2, 1]
RMSE: 0.0027543151820194575 | QLIKE: 0.028456323840714653 | MAE: 0.001460571214929297 | MAPE: 0.14433405037379585


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 117079758


Epoch: 250 | Inputs: 15 | MLP units: [[100, 100], [100, 100]] | Loss: MQLoss() | Scaler Type: robust | Interpolation: linear
Learning Rate: 1 | Block: [3, 3] | Dropout: 0.4 | Pooling: MaxPool1d | Kernel: [4, 2, 1]
RMSE: 0.0027565298195791066 | QLIKE: 0.028170036053204787 | MAE: 0.0014784708920279309 | MAPE: 0.14578261590868513


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 100 | Inputs: 84 | MLP units: [[512, 512], [512, 512]] | Loss: MQLoss() | Scaler Type: robust | Interpolation: nearest
Learning Rate: 0 | Block: [3, 3] | Dropout: 0.3 | Pooling: MaxPool1d | Kernel: [2, 2, 1]
RMSE: 0.0027156843867745723 | QLIKE: 0.027579871738127495 | MAE: 0.0016434552434353086 | MAPE: 0.16028796096309428


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(4)
  m=randrange(4)
  a=randrange(4)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(4)
  a2=randrange(3)
  a3=randrange(2)
  a4=randrange(2)
  model = NHITS(h=1, input_size=n_inputs[i],
                loss=losses[m],
                scaler_type=scaler_type[o],
                learning_rate=learning_rate[n],
                stack_types=['identity','identity'],
                n_blocks=n_blocks[a],
                mlp_units=mlp_units[h],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[n],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=epochs[k],
                dropout_prob_theta = dropout_prob_theta[a1],
                n_pool_kernel_size=n_pool_kernel_size[a2],
                pooling_mode=pooling_mode[a3],
                interpolation_mode=interpolation_mode[a4],
                random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | MLP units: {mlp_units[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | Interpolation: {interpolation_mode[a4]}")
  print(f"Learning Rate: {n} | Block: {n_blocks[a]} | Dropout: {dropout_prob_theta[a1]} | Pooling: {pooling_mode[a3]} | Kernel: {n_pool_kernel_size[a2]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 122760049


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 74900640


Epoch: 250 | Inputs: 21 | MLP units: [[100, 100], [100, 100]] | Loss: MSE() | Scaler Type: robust | Interpolation: nearest
Learning Rate: 3 | Block: [2, 2] | Dropout: 0.2 | Pooling: AvgPool1d | Kernel: [2, 2, 1]
RMSE: 0.0027941861484081744 | QLIKE: 0.02818534019336687 | MAE: 0.0014578721457316334 | MAPE: 0.144969154675579


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 69168416


Epoch: 100 | Inputs: 10 | MLP units: [[100, 100], [100, 100]] | Loss: MSE() | Scaler Type: minmax | Interpolation: nearest
Learning Rate: 0 | Block: [5, 5] | Dropout: 0.3 | Pooling: MaxPool1d | Kernel: [4, 2, 1]
RMSE: 0.0026743734865676667 | QLIKE: 0.025297169437271047 | MAE: 0.0015881318552392729 | MAPE: 0.1622136352513959


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 112684734


Epoch: 750 | Inputs: 15 | MLP units: [[250, 250], [250, 250]] | Loss: MSE() | Scaler Type: standard | Interpolation: nearest
Learning Rate: 2 | Block: [1, 1] | Dropout: 0.4 | Pooling: AvgPool1d | Kernel: [2, 2, 1]
RMSE: 0.0027954460916593806 | QLIKE: 0.02839926006192012 | MAE: 0.0017287775200635896 | MAPE: 0.1751004269445239


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 56520934


Epoch: 450 | Inputs: 15 | MLP units: [[712, 712], [712, 712]] | Loss: MAE() | Scaler Type: minmax | Interpolation: linear
Learning Rate: 0 | Block: [1, 1] | Dropout: 0 | Pooling: MaxPool1d | Kernel: [4, 2, 1]
RMSE: 0.0026605489535124435 | QLIKE: 0.02512516259252608 | MAE: 0.0016316194289556275 | MAPE: 0.15828980433005518


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 250 | Inputs: 5 | MLP units: [[250, 250], [250, 250]] | Loss: MSE() | Scaler Type: robust | Interpolation: nearest
Learning Rate: 2 | Block: [1, 1] | Dropout: 0.3 | Pooling: AvgPool1d | Kernel: [4, 4, 2]
RMSE: 0.002805832538903971 | QLIKE: 0.027073198988484495 | MAE: 0.0014415187149189356 | MAPE: 0.14545941826992032


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(4)
  m=randrange(4)
  a=randrange(4)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(4)
  a2=randrange(3)
  a3=randrange(2)
  a4=randrange(2)
  model = NHITS(h=1, input_size=n_inputs[i],
                loss=losses[m],
                scaler_type=scaler_type[o],
                learning_rate=learning_rate[n],
                stack_types=['identity','identity'],
                n_blocks=n_blocks[a],
                mlp_units=mlp_units[h],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[n],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=epochs[k],
                dropout_prob_theta = dropout_prob_theta[a1],
                n_pool_kernel_size=n_pool_kernel_size[a2],
                pooling_mode=pooling_mode[a3],
                interpolation_mode=interpolation_mode[a4],
                random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | MLP units: {mlp_units[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | Interpolation: {interpolation_mode[a4]}")
  print(f"Learning Rate: {n} | Block: {n_blocks[a]} | Dropout: {dropout_prob_theta[a1]} | Pooling: {pooling_mode[a3]} | Kernel: {n_pool_kernel_size[a2]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 93471300


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 67551000


Epoch: 750 | Inputs: 42 | MLP units: [[100, 100], [100, 100]] | Loss: MSE() | Scaler Type: minmax | Interpolation: nearest
Learning Rate: 3 | Block: [1, 1] | Dropout: 0.2 | Pooling: AvgPool1d | Kernel: [4, 2, 1]
RMSE: 0.002793989506728471 | QLIKE: 0.027975159675650074 | MAE: 0.0014499267370873831 | MAPE: 0.14547318821857133


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 61819626


Epoch: 100 | Inputs: 10 | MLP units: [[100, 100], [100, 100]] | Loss: MSE() | Scaler Type: robust | Interpolation: linear
Learning Rate: 2 | Block: [1, 1] | Dropout: 0.3 | Pooling: MaxPool1d | Kernel: [4, 4, 2]
RMSE: 0.002799294344170138 | QLIKE: 0.026820564258682116 | MAE: 0.0014657116693694533 | MAPE: 0.1483163187835536


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 60152181


Epoch: 650 | Inputs: 3 | MLP units: [[250, 250], [250, 250]] | Loss: MSE() | Scaler Type: minmax | Interpolation: linear
Learning Rate: 2 | Block: [1, 1] | Dropout: 0.4 | Pooling: AvgPool1d | Kernel: [4, 2, 1]
RMSE: 0.002948104578819801 | QLIKE: 0.02681749166677684 | MAE: 0.0016516558230418796 | MAPE: 0.16866123931795365


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 57638957


Epoch: 750 | Inputs: 21 | MLP units: [[250, 250], [250, 250]] | Loss: MSE() | Scaler Type: minmax | Interpolation: linear
Learning Rate: 2 | Block: [1, 1] | Dropout: 0.3 | Pooling: MaxPool1d | Kernel: [4, 4, 2]
RMSE: 0.002682316549960416 | QLIKE: 0.02671892480674497 | MAE: 0.0015642385595562899 | MAPE: 0.1559671433664745


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 750 | Inputs: 5 | MLP units: [[100, 100], [100, 100]] | Loss: DistributionLoss() | Scaler Type: robust | Interpolation: nearest
Learning Rate: 3 | Block: [1, 1] | Dropout: 0.2 | Pooling: MaxPool1d | Kernel: [4, 2, 1]
RMSE: 0.0028061639243976986 | QLIKE: 0.027259903124337806 | MAE: 0.0014395587809333161 | MAPE: 0.14457959604807893


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(4)
  m=randrange(4)
  a=randrange(4)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(4)
  a2=randrange(3)
  a3=randrange(2)
  a4=randrange(2)
  model = NHITS(h=1, input_size=n_inputs[i],
                loss=losses[m],
                scaler_type=scaler_type[o],
                learning_rate=learning_rate[n],
                stack_types=['identity','identity'],
                n_blocks=n_blocks[a],
                mlp_units=mlp_units[h],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[n],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=epochs[k],
                dropout_prob_theta = dropout_prob_theta[a1],
                n_pool_kernel_size=n_pool_kernel_size[a2],
                pooling_mode=pooling_mode[a3],
                interpolation_mode=interpolation_mode[a4],
                random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | MLP units: {mlp_units[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | Interpolation: {interpolation_mode[a4]}")
  print(f"Learning Rate: {n} | Block: {n_blocks[a]} | Dropout: {dropout_prob_theta[a1]} | Pooling: {pooling_mode[a3]} | Kernel: {n_pool_kernel_size[a2]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 58832951


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 105334351


Epoch: 750 | Inputs: 10 | MLP units: [[512, 512], [512, 512]] | Loss: MAE() | Scaler Type: minmax | Interpolation: linear
Learning Rate: 1 | Block: [1, 1] | Dropout: 0.4 | Pooling: AvgPool1d | Kernel: [4, 4, 2]
RMSE: 0.0027742324016831682 | QLIKE: 0.029480907152121046 | MAE: 0.0015744596151699326 | MAPE: 0.15487231434695145


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 23268987


Epoch: 150 | Inputs: 15 | MLP units: [[712, 712], [712, 712]] | Loss: MSE() | Scaler Type: standard | Interpolation: nearest
Learning Rate: 1 | Block: [1, 1] | Dropout: 0.2 | Pooling: AvgPool1d | Kernel: [2, 2, 1]
RMSE: 0.0028127924696157614 | QLIKE: 0.02851340648766623 | MAE: 0.0017445517815600107 | MAPE: 0.17839207947661914


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 17266149


Epoch: 650 | Inputs: 42 | MLP units: [[100, 100], [100, 100]] | Loss: MSE() | Scaler Type: standard | Interpolation: nearest
Learning Rate: 0 | Block: [5, 5] | Dropout: 0 | Pooling: AvgPool1d | Kernel: [2, 2, 1]
RMSE: 0.00276774394230702 | QLIKE: 0.02844941101263773 | MAE: 0.0016776361181663987 | MAPE: 0.16867751394753766


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 9445415


Epoch: 650 | Inputs: 84 | MLP units: [[100, 100], [100, 100]] | Loss: MAE() | Scaler Type: standard | Interpolation: linear
Learning Rate: 1 | Block: [1, 1] | Dropout: 0.4 | Pooling: MaxPool1d | Kernel: [2, 2, 1]
RMSE: 0.0027434541049679443 | QLIKE: 0.027225711964674992 | MAE: 0.001454334020059859 | MAPE: 0.144667803757887


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 650 | Inputs: 21 | MLP units: [[250, 250], [250, 250]] | Loss: MAE() | Scaler Type: standard | Interpolation: linear
Learning Rate: 3 | Block: [3, 3] | Dropout: 0.4 | Pooling: AvgPool1d | Kernel: [4, 2, 1]
RMSE: 0.00272945627831789 | QLIKE: 0.027006931483143988 | MAE: 0.0015017116729681647 | MAPE: 0.1481451512059265


INFO:lightning_fabric.utilities.seed:Seed set to 77483894

Epoch: 150 | Inputs: 21 | MLP units: [[712, 712], [712, 712]] | Loss: DistributionLoss() | Scaler Type: standard | Interpolation: nearest
Learning Rate: 0 | Block: [2, 2] | Dropout: 0 | Pooling: MaxPool1d | Kernel: [4, 2, 1]

RMSE: 0.00261795408412616 | QLIKE: 0.024710279430817574 | MAE: 0.0016030544887469216 | MAPE: 0.15921628282272618

In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = NHITS(h=1, input_size=21,
                loss=losses[3],
                scaler_type='standard',
                learning_rate=learning_rate[0],
                stack_types=['identity','identity'],
                n_blocks=[2,2],
                mlp_units= [[712, 712], [712, 712]],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[0],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=150,
                dropout_prob_theta = 0,
                n_pool_kernel_size=[4,2,1],
                pooling_mode='MaxPool1d' ,
                interpolation_mode='nearest',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 75404644


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 72860215


Random Seed: 75404644
RMSE: 0.002625901125656391 | QLIKE: 0.02499649111275952 | MAE: 0.0015794420564466912 | MAPE: 0.15274075461963318


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 79114046


Random Seed: 72860215
RMSE: 0.002628088544421531 | QLIKE: 0.024724375016896654 | MAE: 0.0015843864383824973 | MAPE: 0.15346692751607957


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 13540218


Random Seed: 79114046
RMSE: 0.0026506585360304828 | QLIKE: 0.02556577661853954 | MAE: 0.001611434854772691 | MAPE: 0.15650325983915606


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 60225863


Random Seed: 13540218
RMSE: 0.002720877322698003 | QLIKE: 0.026592157634466033 | MAE: 0.0016456189308465849 | MAPE: 0.1570686905818709


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 60225863
RMSE: 0.0026655954642769986 | QLIKE: 0.025489856683519704 | MAE: 0.001593193720650095 | MAPE: 0.15188960218351394


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = NHITS(h=1, input_size=21,
                loss=losses[3],
                scaler_type='standard',
                learning_rate=learning_rate[0],
                stack_types=['identity','identity'],
                n_blocks=[2,2],
                mlp_units= [[712, 712], [712, 712]],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[0],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=150,
                dropout_prob_theta = 0,
                n_pool_kernel_size=[4,2,1],
                pooling_mode='MaxPool1d' ,
                interpolation_mode='nearest',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 123422336


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 62926154


Random Seed: 123422336
RMSE: 0.0025974658860156434 | QLIKE: 0.024163837094989808 | MAE: 0.001566486784543027 | MAPE: 0.15437820574502403


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 61474249


Random Seed: 62926154
RMSE: 0.002616770691001289 | QLIKE: 0.024842768064280783 | MAE: 0.0015673897602905536 | MAPE: 0.1522039094517984


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 30588198


Random Seed: 61474249
RMSE: 0.00263279083090959 | QLIKE: 0.024687069769389957 | MAE: 0.0015969574538168777 | MAPE: 0.15539687999407478


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 30530965


Random Seed: 30588198
RMSE: 0.002628919251762875 | QLIKE: 0.025041811810197656 | MAE: 0.0015828990145410212 | MAPE: 0.15293260113820883


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 30530965
RMSE: 0.0026524392654564814 | QLIKE: 0.02519682319643343 | MAE: 0.0015939968525946477 | MAPE: 0.15299752887304482


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = NHITS(h=1, input_size=21,
                loss=losses[3],
                scaler_type='standard',
                learning_rate=learning_rate[0],
                stack_types=['identity','identity'],
                n_blocks=[2,2],
                mlp_units= [[712, 712], [712, 712]],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[0],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=150,
                dropout_prob_theta = 0,
                n_pool_kernel_size=[4,2,1],
                pooling_mode='MaxPool1d' ,
                interpolation_mode='nearest',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 100965809


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 109942366


Random Seed: 100965809
RMSE: 0.0026772034624064672 | QLIKE: 0.025571562371054957 | MAE: 0.0016150434060312833 | MAPE: 0.15496457912162853


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 4585948


Random Seed: 109942366
RMSE: 0.0026545988165036444 | QLIKE: 0.02526189610151371 | MAE: 0.0016114513025781109 | MAPE: 0.1557024506744904


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 121446572


Random Seed: 4585948
RMSE: 0.002700868215297582 | QLIKE: 0.02633833772521769 | MAE: 0.0016365876778800142 | MAPE: 0.15735006409863703


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 85634168


Random Seed: 121446572
RMSE: 0.0025994627712884696 | QLIKE: 0.02435428992997895 | MAE: 0.0015567010830778746 | MAPE: 0.15235582452947602


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 85634168
RMSE: 0.002735827403816384 | QLIKE: 0.026847192707995446 | MAE: 0.0016339817088380948 | MAPE: 0.1540480289438204


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = NHITS(h=1, input_size=21,
                loss=losses[3],
                scaler_type='standard',
                learning_rate=learning_rate[0],
                stack_types=['identity','identity'],
                n_blocks=[2,2],
                mlp_units= [[712, 712], [712, 712]],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[0],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=150,
                dropout_prob_theta = 0,
                n_pool_kernel_size=[4,2,1],
                pooling_mode='MaxPool1d' ,
                interpolation_mode='nearest',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "NHITS-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["NHITS"].values
  else:
    Y_hat=forecasts["NHITS-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 63738212


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 52479872


Random Seed: 63738212
RMSE: 0.002633096642759069 | QLIKE: 0.02512032118324124 | MAE: 0.0015815746120501292 | MAPE: 0.15203886518234472


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 71798678


Random Seed: 52479872
RMSE: 0.002661940901013348 | QLIKE: 0.026999085194257918 | MAE: 0.0015761826134019178 | MAPE: 0.1530855083140498


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 128848235


Random Seed: 71798678
RMSE: 0.0026610061048751438 | QLIKE: 0.02568896130604215 | MAE: 0.0016123923072282628 | MAPE: 0.15504790097064342


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 75075575


Random Seed: 128848235
RMSE: 0.0026693529043606504 | QLIKE: 0.025874387478498104 | MAE: 0.0016211522069902807 | MAPE: 0.15565876051538236


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 75075575
RMSE: 0.0026600858096814625 | QLIKE: 0.025176456202993296 | MAE: 0.0016388522072901474 | MAPE: 0.1609501859245695


# Test

In [ ]:
test_length=int(len(df['NKE'])*0.3)
df1=df.unstack().reset_index(name="Value").rename(columns={"level_0":"unique_id", "Date":"ds", "Value":"y"})

In [ ]:
model = NHITS(h=1, input_size=21,
                loss=losses[3],
                scaler_type='standard',
                learning_rate=learning_rate[0],
                stack_types=['identity','identity'],
                n_blocks=[2,2],
                mlp_units= [[712, 712], [712, 712]],
                windows_batch_size=32,
                num_lr_decays=num_lr_decays[0],
                val_check_steps=500,
                stat_exog_list=['AMT', 'COP', 'CVX', 'AXP', 'BAC', 'C', 'CB', 'CME', 'GS', 'JPM', 'MS',
       'PNC', 'USB', 'WFC', 'BA', 'CAT', 'GE', 'HON', 'LMT', 'MMM', 'UNP',
       'UPS', 'AAPL', 'ACN', 'ADBE', 'ADP', 'CMCSA', 'CRM', 'CSCO', 'DIS',
       'FIS', 'FISV', 'GOOG', 'IBM', 'INTC', 'INTU', 'MA', 'MSFT', 'NFLX',
       'NVDA', 'ORCL', 'QCOM', 'T', 'TXN', 'VZ', 'COST', 'KO', 'MO', 'PEP',
       'PG', 'WMT', 'D', 'DUK', 'SO', 'ABT', 'AMGN', 'BDX', 'BMY', 'BSX', 'CI',
       'CVS', 'DHR', 'GILD', 'ISRG', 'JNJ', 'LLY', 'MDT', 'MRK', 'PFE', 'SYK',
       'TMO', 'UNH', 'AMZN', 'HD', 'LOW', 'MCD', 'NKE', 'SBUX', 'TGT', 'TJX'],
                max_steps=150,
                dropout_prob_theta = 0,
                n_pool_kernel_size=[4,2,1],
                pooling_mode='MaxPool1d' ,
                interpolation_mode='nearest',
              random_seed=123422336)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=1,static_df=df_static,n_windows=None, test_size=test_length-test_length%1,step_size=1)
forecasts = forecasts.dropna()
if "NHITS-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["NHITS"].values
else:
  Y_hat=forecasts["NHITS-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 123422336


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.004007353152580852 | QLIKE: 0.03299555886615222 | MAE: 0.0024607141400636353 | MAPE: 0.1840706474505275


In [ ]:
if "NHITS-median" not in list(forecasts.columns.values):
  df2=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")
  df2=df2.pivot(columns="unique_id")
else:
  df2=forecasts[["ds","unique_id","NHITS-median"]].set_index("ds")
  df2=df2.pivot(columns="unique_id")

In [ ]:
df2.to_excel("NHITS forecasts.xlsx")